### ISSUE TITLE
Plan out how to refactor the service updater

### ISSUE THREAD

**Issue by christophermaier** [habitat-sh/habitat#7389]
The code in the service updater is due for a refresh. 

- election code for rolling updates is hard to reason about and intertwined in general downloading code
- the difference between rolling update and at-once update is both small and confusing
- multiple threads are used to manage updates (one per service) in ways that can suffer from worst-case performance in certain corner-cases
- we can write asynchronous code easily in Rust now, so we should take advantage of that. Can updates be folded into the general lifecycle of a service directly?

All this will be a major undertaking and should be investigated further to come up with a more concrete plan for how to move forward.

**Comment by christophermaier** [comment]
A specific issue that is effectively addressed by #7390 for `at-once` updating, but could still be an issue for rolling updates in the current code appears when an update takes longer to download than the service stays up. For example, a service whose package is 100MB in size and, due to a bug in its `run` hook, exits with a failure as soon as it starts. 

In such a case, every restart of the service will trigger a new service updater thread, which will download the new 100MB package. Because of how these threads are currently structured, they can't be killed while they are downloading a package. The threads and their downloads do not share any information in common, so one thread can't take advantage of anything the others have downloaded yet. Thus, you end up downloading the same large file multiple times, until all threads have succeeded (new threads stop being spawned as soon as the new, presumably non-buggy, version is downloaded and running).

The end result of this is that you can end up saturating your network connection. Depending on the size of the package being downloaded and the available bandwidth, this can be a significant problem.

Part of the issue is that this code predates futures, and the HTTP client code is currently blocking (though that is being addressed in #7362), making it tricky to kill. Another issue (which may go away with refactoring) is that these threads / tasks are spawned assuming that they are the only one present for a given service, which allows them to pile up in this particular scenario. 

In one way, it would be good to completely kill these tasks as soon as the corresponding service shuts down, as that's less bookkeeping to track. On the other hand, if one currently-downloading updater were allowed to continue through restarts, it would allow users to fix the situation by promoting a new package; eventually, the package would download (and only once), and the new service would start. If an in-progress download were killed outright, however, the only remediation would be for an operator to manually install a working package.

That has some nice usability features; if we do that, we should remove the delayed start to update checking that was introduced as a stop-gap measure in #7390 for `at-once` updates.

**Comment by davidMcneil** [comment]
In my opinion, the "best" approach to cleaning up this code is to completely migrate to a future-per-service architecture opposed to the current event-loop-tick architecture.

The architecture I imagine is:
- the main loop spawns a service future when a new service is loaded. This spawned future acts as an ["actor"](https://en.wikipedia.org/wiki/Actor_model)
- the main loop sends events to the "actor" through a channel. Events would include things such as:
  - shutting down
  - restart
  - update
  - detach
  -...
- the actor would have a well-defined set of states represented as an enum. States would include things such as: 
  - initializing
  - running
  - stopping
  - updating
  - detached
  -...
- the actor could communicate events back out to the main loop via another channel if needed

This state machine representation of each service could significantly simplify the update logic as well as other parts of the code (ie hook execution, reasoning about sequences of events within a service, ...). It also would make the network issue a moot point. Updates would be a core part of the lifecycle of each service.

Unfortunately, this proposal only further opens up the refactoring can of worms and probably requires an exploratory issue to start playing with the code and see how this sort of design would work.

**Comment by davidMcneil** [comment]
Issue #7420 captures the work to convert to a future-per-service architecture. Assuming that work happens, I would like to pursue an architecture that completely decouples the actual update from determining when/what to update and the election process.

# Service Actor
The service actor should have an `update` method/message that takes a package ident. When this method is called the service actor checks if the ident is available locally, if it is it updates (no network request needed). If it is not it fetches the package from builder and then updates.

# Service Updater
A global service updater is responsible for orchestrating rolling updates (much like the architecture today). But this updater is only responsible for listening to the gossip protocol and commanding the service actor to update not performing the update itself.

# At-Once Updates
I am not sure if it is better to detect an at-once update as part of the service actor or with the service updater. I lean towards putting the logic in the service updater to keep the detection logic together.

# What version should we update to?
Currently, we simply look for a newer version to trigger an update (I think?). Would it be worth allowing more controlled updates that take into account timestamp, patch, minor, and major version information. 

We could use the package ident stored in the spec file to determine what sort of update is allowed. For example `core/redis/1.3` would update for any increase in the timestamp or patch fields but not for changes to the minor or major fields. This is probably a very substantial change. We dont allow partial version information in a package ident and dont enforce the format of the version.

**Comment by davidMcneil** [comment]
Implementation issue #7427
